app/services/emailAuth.ts

```diff
@@ -0,0 +1,70 @@
+import {
+  sendSignInLinkToEmail,
+  signInWithEmailLink,
+  signOut,
+} from 'firebase/auth';
+import { auth, actionCodeSettings } from '../config/firebase';
+import userDataSanitizationSchema from '../schemas/userDataSanitizationSchema';
+import { storeUserDataIfNew } from './firestore';
+
+/**
+ * Sends authentication link to user email and saves user email address to local storage.
+ * @param {string} email - user's email address
+ */
+export const sendEmailLink = (email: string) => {
+  sendSignInLinkToEmail(auth, email, actionCodeSettings)
+    .then(() => {
+      window.localStorage.setItem('emailForSignIn', email);
+    })
+    .catch((error) => {
+      console.error(
+        'Error occurred while sending email authentication link. Please try again.'
+      );
+    });
+};
+
+/**
+ * Authenticates and signs in user following redirect. Then saves their form data to database.
+ * @param {email} email - User's email address. Must be passed or user will be asked to provide it.
+ */
+export const authenticateAndSaveUserDataFromEmailRedirect = (email: string) => {
+  signInWithEmailLink(auth, email, window.location.href)
+    .then(async (result) => {
+      const localStorageData = {
```

Comment:
To continue the discussion in `app/components/main-content/PrelaunchSignUpForm.tsx`, the data retrieval could look something like (if we want to be double secured we can add fallbacks to empty strings):

```
const localStorageData = {
    email: email,
    name: window.localStorage.getItem('userName'),
    interests: window.localStorage.getItem('userInterest') || '',
    source: window.localStorage.getItem('userSource') || '',
}
```

The Zod sanitization will parse and sanitize each input field (including converting the interests string to an array) before it is written to the database so it may be unnecessary to reduce the user data object before sanitization.

Reply:
That's right! I like your solution, very clean!